financial-templates-lib/price-feed/BalancerPriceFeed.js

```diff
@@ -0,0 +1,79 @@
+const assert = require("assert");
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+const { BlockHistory, PriceHistory } = require("./utils");
+
+// Gets balancer prices as well as historical prices
+class BalancerPriceFeed extends PriceFeedInterface {
+  constructor(logger, web3, getTime, abi, address, tokenIn, tokenOut, lookback) {
+    assert(tokenIn, "BalancerPriceFeed requires tokenIn");
+    assert(tokenOut, "BalancerPriceFeed requires tokenOut");
+    assert(lookback, "BalancerPriceFeed requires lookback");
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+    this.getTime = getTime;
+
+    this.contract = new web3.eth.Contract(abi, address);
+    this.currentPrice = null;
+    this.lastUpdateTime = null;
+    this.tokenIn = tokenIn;
+    this.tokenOut = tokenOut;
+    this.lookback = lookback;
+    // Provide a getblock function which returns the latest value if no number provided.
+    this.blockHistory = BlockHistory(number => web3.eth.getBlock(number >= 0 ? number : "latest"));
+
+    // Add a callback to get price, if an error is thrown returns undefined
+    this.priceHistory = PriceHistory(async number => {
+      try {
+        const result = await this.contract.methods.getSpotPriceSansFee(this.tokenIn, this.tokenOut).call(number);
+        return result;
+      } catch (err) {
+        this.logger.warn({
+          at: "BalancerPriceFeed.getPrice",
+          message: err.message,
+          number
+        });
+      }
+    });
+  }
+  getHistoricalPrice(time) {
+    // normally would bubble up errors, but this is not supposed to throw
+    try {
+      const block = this.blockHistory.getClosestTime(time);
+      return this.priceHistory.get(block.timestamp);
+    } catch (err) {
+      // this can throw an error if no price is found, but lets return null to copy uniswap price feed
+      return null;
+    }
+  }
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+  getCurrentPrice() {
+    // normally would bubble up errors, but this is not supposed to throw
+    try {
+      return this.priceHistory.currentPrice();
```

Comment:
Do we need to do any post-processing of the prices? In what form are the prices that balancer returns? How are they decimalized? Note: this isn't to tackle the 18 decimal issue -- we just want to make sure it's consistent with the other price feeds for our use cases.

Reply:
this is a concern of mine. its not documented, and i will probably need to go to source or make calls to contract to deduce.

Reply:
ive looked into this more, balancer source is really really nice. they provided the formula they use 

```
    /**********************************************************************************************
    // calcSpotPrice                                                                             //
    // sP = spotPrice                                                                            //
    // bI = tokenBalanceIn                ( bI / wI )         1                                  //
    // bO = tokenBalanceOut         sP =  -----------  *  ----------                             //
    // wI = tokenWeightIn                 ( bO / wO )     ( 1 - sF )                             //
    // wO = tokenWeightOut                                                                       //
    // sF = swapFee                                                                              //
    **********************************************************************************************/
```
This is our uniswap price calculation 
![image](https://user-images.githubusercontent.com/4429761/90282834-11217c80-de3d-11ea-9de9-01b4444745f0.png)

This looks essentially the same (with a few extra balancer vars), meaning I shouldnt need to do any conversions after the price call, though its unclear to me which should be numerator or denominator. currently tokenin (numerator) is the emp synthetic and token out (denomintor) is the liquidity token.

Reply:
Perfect. Yeah, I think all of that looks right to me.